amethyst_rendy/shaders/fragment/header/ltc.glsl

```diff
@@ -0,0 +1,540 @@
+        float Theta = atan(D*sqrt(Discriminant), -D_d)/3.0;
+
+        float x_1d = 2.0*sqrt(-C_d)*cos(Theta);
+        float x_3d = 2.0*sqrt(-C_d)*cos(Theta + (2.0/3.0)*PI);
+
+        float xs;
+        if (x_1d + x_3d < 2.0*C)
+            xs = x_1d;
+        else
+            xs = x_3d;
+
+        xsc = vec2(-D, xs + C);
+    }
+
+    float E =  xlc.y*xsc.y;
+    float F = -xlc.x*xsc.y - xlc.y*xsc.x;
+    float G =  xlc.x*xsc.x;
+
+    vec2 xmc = vec2(C*F - B*G, -B*F + C*E);
+
+    vec3 Root = vec3(xsc.x/xsc.y, xmc.x/xmc.y, xlc.x/xlc.y);
+
+    if (Root.x < Root.y && Root.x < Root.z)
+        Root.xyz = Root.yxz;
+    else if (Root.z < Root.x && Root.z < Root.y)
+        Root.xyz = Root.xzy;
+
+    return Root;
+}
+
+
+/*
+ * Get intensity of light from the arealight given by `points` at the point `P`
+ * with normal `N` when viewed from direction `P`.
+ * @param N Normal
+ * @param V View Direction
+ * @param P Vertex Position
+ * @param Minv Matrix to transform from BRDF distribution to clamped cosine distribution
+ * @param points Light quad vertices
+ * @param twoSided Whether the light is two sided
+ */
+vec3 ltc_evaluate_ellipse(
+    vec3 N, 
+    vec3 V, 
+    vec3 P, 
+    float NdotV,
+    mat3 Minv, 
+    vec3 points[4], 
+    bool two_sided,
+    bool sphere
+)
+{
+
+
+    if (sphere) {
+        // Get center of our quad.
+        vec3 Cr;
+        Cr = (points[0] + points[2])/2.0;
+        
+        vec3 Cv0 = points[0] - Cr;
+        vec3 Cv1 = points[1] - Cr;
+        vec3 Cv2 = points[2] - Cr;
+        
+        vec3 Cn = normalize(cross(points[0]-points[2], points[1]- points[3]));
+        vec3 Cp = normalize(P - Cr);
+        if (Cn != Cp) {
+            float costheta = dot(-Cn, Cp);
+            vec3 rotAxis = cross(Cn, Cp);
+            mat3 Z = rotationMatrix3(rotAxis, costheta);
+            // rotate points around center to change the plane to be perpendicular to the vector from center to fragment.
+            points[0] = Cr + (Z * Cv0);
+            points[1] = Cr + (Z * Cv1);
+            points[2] = Cr + (Z * Cv2);
+        }
+    }
+
+    // construct orthonormal basis around N
+    vec3 T1, T2;
+    T1 = normalize(V - N*NdotV);
+    T2 = cross(N, T1);
+
+    // rotate area light in (T1, T2, N) basis
+    mat3 R = transpose(mat3(T1, T2, N));
+
+    // polygon (allocate 5 vertices for clipping)
+    vec3 L_[3];
+    
+    L_[0] = R * (points[0] - P);
+    L_[1] = R * (points[1] - P);
+    L_[2] = R * (points[2] - P);
+
+
+    // init ellipse
+    vec3 C  = 0.5 * (L_[0] + L_[2]);
+    vec3 V1 = 0.5 * (L_[1] - L_[2]);
+    vec3 V2 = 0.5 * (L_[1] - L_[0]);
+
+    C  = Minv * C;
+    V1 = Minv * V1;
+    V2 = Minv * V2;
+
+    if(!two_sided && dot(cross(V1, V2), C) < 0.0)
+        return vec3(0.0);
+
+    // compute eigenvectors of ellipse
+    float a, b;
+    float d11 = dot(V1, V1);
+    float d22 = dot(V2, V2);
+    float d12 = dot(V1, V2);
+    if (abs(d12)/sqrt(d11*d22) > 0.0001)
```

Comment:
Same here. I wonder if we can more appropriate handle the divide by 0's this is trying to avoid.

Reply:
The divide by 0's are also causing some issues I need to tackle in the solve_cubic function.
Would be interesting if any AAA games already used this approach.